great_expectations/datasource/fluent/snowflake_datasource.py

```diff
@@ -531,8 +531,15 @@ def get_engine(self) -> sqlalchemy.Engine:
                 self._cached_connection_string = self.connection_string
         return self._engine
 
-    def _build_engine_with_connect_args(self, **kwargs) -> sqlalchemy.Engine:
-        connect_args = self._get_connect_args()
-        connect_args.update(kwargs)
-        url = URL(**connect_args)
-        return sa.create_engine(url)
+    def _build_engine_with_connect_args(
+        self, connect_args: dict[str, Any] | None = None, **kwargs
+    ) -> sqlalchemy.Engine:
+        url_args = self._get_url_args()
+        url_args.update(kwargs)
+        if connect_args:
+            if connect_args.get("private_key"):
+                url_args.pop(  # TODO: update models + validation to handle this
+                    "password", None
+                )
+        url = URL(**url_args)
+        return sa.create_engine(url, connect_args=connect_args)
```

Comment:
With this minimal change users can connect to snowflake by providng a dummy password.

```python

import pathlib
import great_expectations as gx

from cryptography.hazmat.backends import default_backend
from cryptography.hazmat.primitives.asymmetric import rsa
from cryptography.hazmat.primitives.asymmetric import dsa
from cryptography.hazmat.primitives import serialization


ephem_ctx = gx.get_context(mode="ephemeral")

PRIVATE_KEY_FILE = pathlib.Path("rsa_key.p8").resolve(strict=True)

p_key = serialization.load_pem_private_key(
        PRIVATE_KEY_FILE.read_bytes(),
        password=b"my_password",
        backend=default_backend()
    )

pkb = p_key.private_bytes(
    encoding=serialization.Encoding.DER,
    format=serialization.PrivateFormat.PKCS8,
    encryption_algorithm=serialization.NoEncryption())

connect_args = {"private_key": pkb}

connection_details={
    "account": "my_account.us-east-1",
    "user": "my_user",
    "role": "my_role",
    "password": "DUMMY_VALUE", # modified code will ignore this
    "warehouse": "my_wh",
    "database": "my_db",
    "schema": "my_schema"
}

sf_ds = ephem_ctx.sources.add_snowflake(
    "my_ds",
    connection_string=connection_details,
    kwargs={"connect_args": connect_args}
)
```



We should update the `SnowflakeDatasource` model to accept `private_key` and/or `connect_args` as a first class parameter. And not require a password in that instance.

Reply:
With some minor additional changes it could look more like this...
```python
...
connection_details={
    "account": "my_account.us-east-1",
    "user": "my_user",
    "role": "my_role",
    "private_key": pkb,
    "warehouse": "my_wh",
    "database": "my_db",
    "schema": "my_schema"
}

sf_ds = ephem_ctx.sources.add_snowflake(
    "my_ds",
    connection_string=connection_details,
)
```